snapshotter/devmapper/metadata.go

```diff
@@ -0,0 +1,312 @@
+			return err
+		}
+
+		if _, err := tx.CreateBucketIfNotExists(deviceIDBucketName); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+// AddDevice saves device info to database.
+// Callback should be used to acquire device ID or to rollback transaction in case of error.
+func (m *PoolMetadata) AddDevice(ctx context.Context, info *DeviceInfo, fn DeviceIDCallback) error {
+	return m.db.Update(func(tx *bolt.Tx) error {
+		devicesBucket := tx.Bucket(devicesBucketName)
+
+		// Make sure device name is unique
+		if err := getObject(devicesBucket, info.Name, nil); err == nil {
+			return ErrAlreadyExists
+		}
+
+		// Find next available device ID
+		deviceID, err := getNextDeviceID(tx)
+		if err != nil {
+			return err
+		}
+
+		if err := fn(deviceID); err != nil {
+			return err
+		}
+
+		info.DeviceID = deviceID
+
+		if err := putObject(devicesBucket, info.Name, info, false); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func getNextDeviceID(tx *bolt.Tx) (int, error) {
+	bucket := tx.Bucket(deviceIDBucketName)
+	cursor := bucket.Cursor()
+
+	// Check if any device id can be reused.
+	// Bolt stores its keys in byte-sorted order within a bucket.
+	// This makes sequential iteration extremely fast.
+	for key, taken := cursor.First(); key != nil; key, taken = cursor.Next() {
+		isFree := taken[0] == deviceFree
+		if isFree {
+			id, err := strconv.Atoi(string(key))
+			if err != nil {
+				return 0, err
+			}
+
+			if err := markDeviceID(tx, id, deviceTaken); err != nil {
+				return 0, err
+			}
+
+			return id, nil
+		}
+	}
+
+	// Try allocate new device ID
+	seq, err := bucket.NextSequence()
+	if err != nil {
+		return 0, err
+	}
+
+	if seq >= maxDeviceID {
+		return 0, errors.Errorf("couldn't find free device key")
+	}
+
+	id := int(seq)
+	if err := markDeviceID(tx, id, deviceTaken); err != nil {
+		return 0, err
+	}
+
+	return id, nil
+}
+
+func markDeviceID(tx *bolt.Tx, deviceID int, state byte) error {
+	var (
+		bucket = tx.Bucket(deviceIDBucketName)
+		key    = strconv.Itoa(deviceID)
+		value  = []byte{state}
+	)
+
+	if err := bucket.Put([]byte(key), value); err != nil {
+		return errors.Wrapf(err, "failed to free device id %q", key)
+	}
+
+	return nil
+}
+
+// UpdateDevice updates device info in metadata store.
+// Callback should be used to modify device properties or to rollback update transaction.
+// Name and Device ID couldn't be changed and will be ignored.
+func (m *PoolMetadata) UpdateDevice(ctx context.Context, name string, fn DeviceInfoCallback) error {
+	return m.db.Update(func(tx *bolt.Tx) error {
+		var (
+			device = &DeviceInfo{}
+			bucket = tx.Bucket(devicesBucketName)
+		)
+
+		if err := getObject(bucket, name, device); err != nil {
+			return err
+		}
+
+		// Don't allow changing these values, keep things in sync with devmapper
+		name := device.Name
+		devID := device.DeviceID
+
+		if err := fn(device); err != nil {
+			return err
+		}
+
+		device.Name = name
```

Comment:
I'm not sure I want to overcomplicate this. metadata is not going to be used outside `devmapper` package and there are no cases when we want to update name or device id.
Also I've added a comment which states that these fields are ignored.

Reply:
If there are **no** cases, why not just remove the cached values and pass in `device`? Otherwise I vote for returning an error. I'd much rather have less magic or potential for less confusion.

Reply:
I'd like to see an explicit error instead of silently ignoring the change.